atef/check.py

```diff
@@ -0,0 +1,707 @@
+from __future__ import annotations
+
+import enum
+import logging
+from dataclasses import dataclass, field
+from typing import (Any, Callable, ClassVar, Dict, Generator, List, Optional,
+                    Sequence, Tuple, Type, Union)
+
+import numpy as np
+import ophyd
+
+from . import serialization
+
+Number = Union[int, float]
+
+logger = logging.getLogger(__name__)
+
+
+class Severity(enum.IntEnum):
+    success = 0
+    warning = 1
+    error = 2
+    internal_error = 3
+
+
+class ComparisonException(Exception):
+    """Raise this exception to exit a comparator and set severity."""
+    severity = Severity.success
+
+
+class ComparisonError(ComparisonException):
+    """Raise this exception to error out in a comparator."""
+    severity = Severity.error
+
+
+class ComparisonWarning(ComparisonException):
+    """Raise this exception to warn in a comparator."""
+    severity = Severity.warning
+
+
+class ReduceMethod(str, enum.Enum):
+    average = "average"
+    median = "median"
+    sum = "sum"
+    min = "min"
+    max = "max"
+    std = "std"
+
+    @property
+    def method(self) -> Callable:
+        return {
+            ReduceMethod.average: np.average,
+            ReduceMethod.median: np.median,
+            ReduceMethod.sum: np.sum,
+            ReduceMethod.min: np.min,
+            ReduceMethod.max: np.max,
+            ReduceMethod.std: np.std,
+        }[self]
```

Comment:
I've never considered this enum pattern before, it seems like it could be useful